File: leetcode/questions/863/python/863.py
from collections import defaultdict
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        def build_graph():
            graph = defaultdict(list)
            stack = [(root, None)]

            while stack:
                node, parent = stack.pop()

                if not node:
                    continue

                if parent:
                    graph[node.val].append(parent.val)
                    graph[parent.val].append(node.val)

                stack.append((node.left, node))
                stack.append((node.right, node))

            return graph

        graph = build_graph()

        def dfs(root):
            stack = [(root, 0)]
            marked = set()
            ans = []

            while stack:
                node, dist = stack.pop()

                if node in marked:
                    continue

                marked.add(node)

                if dist == k:
                    ans.append(node)
                    continue

                for nxt in graph[node]:
                    stack.append((nxt, dist + 1))

            return ans

        return dfs(target.val)

File: leetcode/questions/863/java/863.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Solution {
    List<Integer> answer = new ArrayList<>();
    HashMap<Integer, List<Integer>> graph = new HashMap<>();
    HashSet<Integer> visited = new HashSet<>();
    int k;

    private void createGraph(TreeNode root) {
        ArrayDeque<TreeNode[]> stack = new ArrayDeque<>();
        stack.add(new TreeNode[] { root, null });

        while (!stack.isEmpty()) {
            TreeNode[] nodes = stack.pollLast();
            TreeNode cur = nodes[0], parent = nodes[1];

            if (cur == null) {
                continue;
            }

            if (parent != null) {
                this.graph.computeIfAbsent(cur.val, x -> new ArrayList<>()).add(parent.val);
                this.graph.computeIfAbsent(parent.val, x -> new ArrayList<>()).add(cur.val);
            }

            stack.addLast(new TreeNode[] { cur.left, cur });
            stack.addLast(new TreeNode[] { cur.right, cur });
        }
    }

    private void dfs(int root, int distance) {
        if (visited.contains(root)) {
            return;
        }

        visited.add(root);

        if (distance == this.k) {
            this.answer.add(root);
            return;
        }

        this.graph.getOrDefault(root, new ArrayList<>()).forEach(next -> {
            dfs(next, distance + 1);
        });
    }

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        this.k = k;
        createGraph(root);
        dfs(target.val, 0);

        return this.answer;
    }
}